compiler/compiler_test.go

```diff
@@ -119,28 +502,146 @@ func compile(path string, sourceFiles []source, minify bool) ([]byte, error) {
 		},
 	}
 
-	a, err := importContext.Import(path)
-	if err != nil {
-		return nil, err
-	}
-	b, err := renderPackage(a)
+	_, err := importContext.Import(root.PkgPath)
 	if err != nil {
-		return nil, err
+		t.Fatal(`failed to compile:`, err)
 	}
-	return b, nil
+	return archiveCache
 }
 
-func renderPackage(archive *Archive) ([]byte, error) {
+func renderPackage(t *testing.T, archive *Archive, minify bool) []byte {
+	t.Helper()
 	selection := make(map[*Decl]struct{})
 	for _, d := range archive.Declarations {
 		selection[d] = struct{}{}
 	}
 
 	buf := &bytes.Buffer{}
 
-	if err := WritePkgCode(archive, selection, goLinknameSet{}, false, &SourceMapFilter{Writer: buf}); err != nil {
-		return nil, err
+	if err := WritePkgCode(archive, selection, goLinknameSet{}, minify, &SourceMapFilter{Writer: buf}); err != nil {
+		t.Fatal(err)
 	}
 
-	return buf.Bytes(), nil
+	return buf.Bytes()
+}
+
+type selectionTester struct {
+	t            *testing.T
+	mainPkg      *Archive
+	archives     map[string]*Archive
+	packages     []*Archive
+	dceSelection map[*Decl]struct{}
+
+	DeclCode       *selectionCodeTester
+	InitCode       *selectionCodeTester
+	MethodListCode *selectionCodeTester
+}
+
+func declSelection(t *testing.T, sourceFiles []source, auxFiles []source) *selectionTester {
+	t.Helper()
+	root := parseSources(t, sourceFiles, auxFiles)
+	archives := compileProject(t, root, false)
+	mainPkg := archives[root.PkgPath]
+
+	paths := make([]string, 0, len(archives))
+	for path := range archives {
+		paths = append(paths, path)
+	}
+	sort.Strings(paths)
+	packages := make([]*Archive, 0, len(archives)-1)
```

Comment:
Why is the capacity `len(archives)-1`. Considering the array is filled by iterating over `pats`, which has size `len(archives)`, it seems like an odd choice.

Reply:
Oops, that is odd. That was left over from something I tried. I had the code more like `ImportDependencies` where the root package is added to the end of the slice, then in this code was taking the root package off the end. Later I removed that complexity since, although more similar to the real thing, it wasn't needed for the tests.

I thought I had cleaned it all up, oh well, thanks for catching that.